### ISSUE TITLE
Thread Globals vs Broadway vs Cluster

### ISSUE THREAD

**Issue by feamcor** [k2view-academy/K2View-Academy#743]
Fabric 6.5.2

If I wrap `UserCode.(set|get|clear)ThreadGlobals` on LU functions and export them as actors, will they work properly within Broadway flows?
At least within Broadway population flows?
I wouldn't be using them within async flows.

I assume that thread globals are visible only within the current node/JVM instance.
Is that correct?
If so, is there a cluster-wise counterpart?
Would that be (global) globals (persisted in Cassandra)?
If so, can it handle thousands or more globals being set and unset all the time, for a short period of time?

Please advise.
Thanks.

**Comment by feamcor** [comment]
I see that cluster-wise globals cannot be created on-the-fly, unless they are predefined in the project. Is there a way to overcome this restriction?

**Comment by feamcor** [comment]
Bottom-line is that I would like to know if there is any distributed lock mechanism in Fabric.

**Comment by IvanKostyuk** [comment]
For `ThreadGlobals` the key word is **Thread**. The globals are visible within the thread. It is better to create an actor than using LU functions, but they will work properly within Broadway flow. My concern (need to try)  InnerFlowAsync and Parallel on stage - it is not a fork of a running process, it is a new thread from a thread pool.

I would not use `ThreadGlobals` for population flows, but more simple and reliable mechanism as **Session Globals** just call fabric command `set whatEverNameYouLike=WithWhatEverValue`, then use set to read. This one on scope of session. If on configuration level it is allowed to process same order of population in parallel, then there is no impact. It just works.

Async or not, it is safer when an implementation uses what works in all cases.

Yes, ThreadGlobals are visible within the current node/JVM instance and as the name says within the tread - another thread where the global was not set does no have it.

For cluster-wise only Global globals work, but they have to be defined on development time. Unfortunately, only session ones could be set without declaration.

Cassandra can deal with millions of records set/unset and even with TTL, but I'm not aware of that mechanism in Fabric (will be nice to have - like temporary globals - a name is defined in runtime).

I'm looking forward on implementation a reliable distributed functionality such as sequences and locks in Fabric.

**Comment by lionbro1** [comment]
The requirement to set a Global on cluster level although it was not defined in the implementation was captured and will be evaluated for next releases.
